core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -588,7 +588,10 @@ public final NodeChannels openConnection(DiscoveryNode node, ConnectionProfile c
                         }
                     }
                 };
-                nodeChannels = connectToChannels(node, connectionProfile, onClose);
+                nodeChannels = connectToChannels(node, connectionProfile, this::onChannelOpen, onClose);
+                if (!Arrays.stream(nodeChannels.channels).allMatch(this::isOpen)) {
```

Comment:
I this this to be moved to after ` connectionRef.set(nodeChannels);` is called. Otherwise we still have a race condition when things go wrong later on?

Reply:
Yes, I pushed c4ae4c616f113c1a6f9014a26e90e9e6b5ea7e5e.